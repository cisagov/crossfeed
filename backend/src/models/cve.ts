import {
  Entity,
  PrimaryGeneratedColumn,
  Column,
  ManyToMany,
  BaseEntity,
  JoinTable,
  Unique
} from 'typeorm';
import { Cpe } from './cpe';

//TODO: Refactor column names to camelCase to match the rest of the codebase?
@Entity()
@Unique(['name'])
export class Cve extends BaseEntity {
  @PrimaryGeneratedColumn('uuid')
  id: string;

  @Column({ nullable: true })
  name: string;

  @Column({ nullable: true })
  publishedAt: Date;

  @Column({ nullable: true })
  modifiedAt: Date;

  @Column({ nullable: true })
  status: string;

  @Column({ nullable: true })
  description: string;

  @Column({ nullable: true })
  cvssV2Source: string;

  @Column({ nullable: true })
  cvssV2Type: string;

  @Column({ nullable: true })
  cvssV2Version: string;

  @Column({ nullable: true })
  cvssV2VectorString: string;

  @Column({ nullable: true })
  cvssV2BaseScore: string;

  @Column({ nullable: true })
  cvssV2BaseSeverity: string;

  @Column({ nullable: true })
  cvssV2ExploitabilityScore: string;

  @Column({ nullable: true })
  cvssV2ImpactScore: string;

  @Column({ nullable: true })
  cvssV3Source: string;

  @Column({ nullable: true })
  cvssV3Type: string;

  @Column({ nullable: true })
  cvssV3Version: string;

  @Column({ nullable: true })
  cvssV3VectorString: string;

  @Column({ nullable: true })
  cvssV3BaseScore: string;

  @Column({ nullable: true })
  cvssV3BaseSeverity: string;

  @Column({ nullable: true })
  cvssV3ExploitabilityScore: string;

  @Column({ nullable: true })
  cvssV3ImpactScore: string;

  @Column({ nullable: true })
  cvssV4Source: string;

  @Column({ nullable: true })
  cvssV4Type: string;

  @Column({ nullable: true })
  cvssV4Version: string;

  @Column({ nullable: true })
  cvssV4VectorString: string;

  @Column({ nullable: true })
  cvssV4BaseScore: string;

  @Column({ nullable: true })
  cvssV4BaseSeverity: string;

  @Column({ nullable: true })
  cvssV4ExploitabilityScore: string;

  @Column({ nullable: true })
  cvssV4ImpactScore: string;

  @Column('simple-array', { nullable: true })
  weaknesses: string[];

  @Column('simple-array', { nullable: true })
  references: string[];

  @ManyToMany(() => Cpe, (cpe) => cpe.cves, {
    cascade: true
  })
  @JoinTable()
  cpes: Cpe[];
}
