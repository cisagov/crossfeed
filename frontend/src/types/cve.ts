import { ProductInfo } from './product-info';
export interface Cve {
  cve_uid: string;
  cve_name: string | null;
  description: string | null;
  last_modified_date: Date;
  published_date: Date;
  vuln_status: string | null;
  cvss_v2_source: string | null;
  cvss_v2_type: string | null;
  cvss_v2_version: string | null;
  cvss_v2_vector_string: string | null;
  cvss_v2_base_score: string | null;
  cvss_v2_base_severity: string | null;
  cvss_v2_exploitability_score: string | null;
  cvss_v2_impact_score: string | null;
  cvss_v3_source: string | null;
  cvss_v3_type: string | null;
  cvss_v3_version: string | null;
  cvss_v3_vector_string: string | null;
  cvss_v3_base_score: string | null;
  cvss_v3_base_severity: string | null;
  cvss_v3_exploitability_score: string | null;
  cvss_v3_impact_score: string | null;
  cvss_v4_source: string | null;
  cvss_v4_type: string | null;
  cvss_v4_version: string | null;
  cvss_v4_vector_string: string | null;
  cvss_v4_base_score: string | null;
  cvss_v4_base_severity: string | null;
  cvss_v4_exploitability_score: string | null;
  cvss_v4_impact_score: string | null;
  cpe_list: string[] | null;
  reference_urls: string[] | null;
  weaknesses: string[] | null;
  product_info: ProductInfo[];
}
