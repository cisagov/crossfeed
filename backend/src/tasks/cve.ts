import { Domain, connectToDatabase, Vulnerability } from '../models';
import { spawnSync, execSync } from 'child_process';
import { plainToClass } from 'class-transformer';
import { CommandOptions } from './ecs-client';
import * as buffer from 'buffer';
import saveVulnerabilitiesToDb from './helpers/saveVulnerabilitiesToDb';

/**
 * The CVE scan finds vulnerable CVEs affecting domains based on CPEs identified
 */

export const handler = async (commandOptions: CommandOptions) => {
  console.log('Running cve detection globally');

  await connectToDatabase();
  const allDomains = await Domain.find({
    select: ['id', 'name', 'ip', 'webTechnologies'],
    relations: ['services']
  });
  const hostsToCheck: Array<{
    domain: Domain;
    cpes: string[];
  }> = [];
  for (const domain of allDomains) {
    const cpes = new Set<string>();
    for (const tech of domain.webTechnologies) {
      if (tech.cpe && tech.version) cpes.add(tech.cpe + ':' + tech.version);
    }

    for (const service of domain.services) {
      if (
        service.censysMetadata &&
        service.censysMetadata.manufacturer &&
        service.censysMetadata.product &&
        service.censysMetadata.version
      ) {
        // TODO: Improve methods for getting CPEs from Censys
        // See https://www.napier.ac.uk/~/media/worktribe/output-1500093/identifying-vulnerabilities-using-internet-wide-scanning-data.pdf
        // and https://github.com/TheHairyJ/Scout
        cpes.add(
          `cpe:/a:${service.censysMetadata.manufacturer}:${service.censysMetadata.product}:${service.censysMetadata.version}`.toLowerCase()
        );
      }
    }
    if (cpes.size > 0)
      hostsToCheck.push({
        domain: domain,
        cpes: Array.from(cpes)
      });
  }

  spawnSync('nvdsync', ['-cve_feed', 'cve-1.1.json.gz', 'nvd-dump'], {
    stdio: [process.stdin, process.stdout, process.stderr]
  });

  let input = '';
  for (const [index, host] of hostsToCheck.entries()) {
    input += `${index} ${host.cpes.join(',')}\n`;
    console.log(`${index} ${host.cpes.join(',')}`);
  }

  // Should change this to spawnSync
  const res = execSync(
    "cpe2cve -d ' ' -d2 , -o ' ' -o2 , -cpe 2 -e 2 -matches 3 -cve 2 -cvss 4 -cwe 5 nvd-dump/nvdcve-1.1-2*.json.gz",
    { input: input, maxBuffer: buffer.constants.MAX_LENGTH }
  );

  const split = String(res).split('\n');
  const vulnerabilities: Vulnerability[] = [];
  for (const line of split) {
    const parts = line.split(' ');
    if (parts.length < 5) continue;
    const domain = hostsToCheck[parseInt(parts[0])].domain;

    const cvss = parseFloat(parts[3]);
    let severity: string;

    if (cvss === 0) severity = 'None';
    else if (cvss < 4) severity = 'Low';
    else if (cvss < 7) severity = 'Medium';
    else if (cvss < 9) severity = 'High';
    else severity = 'Critical';

    vulnerabilities.push(
      plainToClass(Vulnerability, {
        domain: domain,
        lastSeen: new Date(Date.now()),
        title: parts[1],
        cve: parts[1],
        cwe: parts[4],
        cpe: parts[2],
        cvss,
        severity,
        state: 'open'
      })
    );
  }
  await saveVulnerabilitiesToDb(vulnerabilities);
};
