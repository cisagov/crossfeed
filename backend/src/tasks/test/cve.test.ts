import { handler as cve } from '../cve';
import {
  Organization,
  Domain,
  connectToDatabase,
  Service,
  Vulnerability
} from '../../models';

jest.mock('child_process', () => ({
  spawnSync: () => null,
  execSync: (cmd, { input }) => {
    expect(input).toMatchSnapshot('execSync.input');
    return [
      '0 CVE-2019-10866 cpe:/a:10web:form_maker:1.0.0 9.8 CWE-89',
      '0 CVE-2019-11590 cpe:/a:10web:form_maker:1.0.0 8.8 CWE-352'
    ].join('\n');
  }
}));

const RealDate = Date;

describe('cve', () => {
  beforeAll(async () => {
    await connectToDatabase();
  });
  beforeEach(() => {
    global.Date.now = jest.fn(() => new Date('2019-04-22T10:20:30Z').getTime());
  });

  afterEach(() => {
    global.Date = RealDate;
  });
  test('simple test', async () => {
    const organization = await Organization.create({
      name: 'test-' + Math.random(),
      rootDomains: ['test-' + Math.random()],
      ipBlocks: [],
      isPassive: false
    }).save();
    const name = 'test-' + Math.random();
    const domain = await Domain.create({
      name,
      organization
    }).save();
    const service = await Service.create({
      domain,
      port: 80,
      censysMetadata: {
        manufacturer: '10web',
        product: 'form_maker',
        version: '1.0.0'
      }
    }).save();
    await cve({
      scanId: 'scanId',
      scanName: 'scanName',
      scanTaskId: 'scanTaskId'
    });

    const vulnerabilities = await Vulnerability.find({
      domain
    });
    expect(vulnerabilities.length).toEqual(2);
    for (const vulnerability of vulnerabilities) {
      expect(vulnerability).toMatchSnapshot(
        {
          id: expect.any(String),
          createdAt: expect.any(Date)
        },
        'vulnerability'
      );
    }
  });
});
