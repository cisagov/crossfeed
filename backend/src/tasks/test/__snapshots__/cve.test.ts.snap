// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cve simple test: execSync.input 1`] = `
"0 cpe:/a:10web:form_maker:1.0.0
"
`;

exports[`cve simple test: vulnerability 1`] = `
Object {
  "cpe": "cpe:/a:10web:form_maker:1.0.0",
  "createdAt": Any<Date>,
  "cve": "CVE-2019-10866",
  "cvss": "9.8",
  "cwe": "CWE-89",
  "description": "",
  "id": Any<String>,
  "lastSeen": 2019-04-22T10:20:30.000Z,
  "severity": "Critical",
  "state": "open",
  "title": "CVE-2019-10866",
}
`;

exports[`cve simple test: vulnerability 2`] = `
Object {
  "cpe": "cpe:/a:10web:form_maker:1.0.0",
  "createdAt": Any<Date>,
  "cve": "CVE-2019-11590",
  "cvss": "8.8",
  "cwe": "CWE-352",
  "description": "",
  "id": Any<String>,
  "lastSeen": 2019-04-22T10:20:30.000Z,
  "severity": "High",
  "state": "open",
  "title": "CVE-2019-11590",
}
`;
